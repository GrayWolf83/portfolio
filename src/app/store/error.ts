import { AppDispatch, RootState } from './index'
import { createSlice, PayloadAction } from '@reduxjs/toolkit'

type errorState = {
	error: string | null
}

const initialState: errorState = {
	error: null,
}

const errorSlice = createSlice({
	name: 'errors',
	initialState,
	reducers: {
		loadingError(state, action: PayloadAction<string>) {
			state.error = action.payload
		},
		errorCleared(state) {
			state.error = null
		},
	},
})

const { loadingError, errorCleared } = errorSlice.actions

export const setLoadingError =
	(error: any) => async (dispatch: AppDispatch) => {
		if (error?.response?.data?.message) {
			dispatch(loadingError(error.response.data.message))
		}
	}

export const clearErrors = () => (dispatch: AppDispatch) => {
	dispatch(errorCleared())
}

export const getError = () => (state: RootState) => {
	return state.errors.error
}

export default errorSlice.reducer
