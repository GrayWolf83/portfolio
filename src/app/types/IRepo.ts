export interface IRepo {
	name: string
	full_name: string
	homepage: string
	description: string
	language: string
	created_at: string
}
