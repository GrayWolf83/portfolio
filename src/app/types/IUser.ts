export interface IUser {
	name: string
	avatar_url: string
	login: string
	html_url: string
	blog: string
	location: string
	created_at: string
	public_repos: number
}
