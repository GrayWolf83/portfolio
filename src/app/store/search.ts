import { AppDispatch, RootState } from './index'
import { setLoadingError } from './error'
import { createSlice, PayloadAction } from '@reduxjs/toolkit'
import { IUser } from '../types/IUser'
import userService from '../services/user.service'

type SearchState = {
	entities: IUser[]
	isLoading: boolean
	error: string | null
	dataLoaded: boolean
	page: number
	search: string
	usersCount: number
}

const initialState: SearchState = {
	entities: [],
	isLoading: false,
	error: null,
	dataLoaded: false,
	page: 1,
	search: '',
	usersCount: 0,
}

const searchSlice = createSlice({
	name: 'search',
	initialState,
	reducers: {
		serachLoadingStart: (state) => {
			state.isLoading = true
		},
		serachLoadingEnd: (state) => {
			state.isLoading = false
		},
		searchDataLoaded: (
			state,
			action: PayloadAction<{
				entities: IUser[]
				usersCount: number
				search: string
				page: number
			}>,
		) => {
			state.entities = action.payload.entities
			state.dataLoaded = true
			state.page = action.payload.page
			state.usersCount = action.payload.usersCount
			state.search = action.payload.search
		},
	},
})

const { serachLoadingStart, serachLoadingEnd, searchDataLoaded } =
	searchSlice.actions

export const searchUsersList =
	(search: string, page: number) => async (dispatch: AppDispatch) => {
		dispatch(serachLoadingStart())
		try {
			const payload = await userService.searchUsers(search, page)
			dispatch(
				searchDataLoaded({
					entities: payload.items,
					usersCount: payload.total_count,
					search,
					page,
				}),
			)
		} catch (error: any) {
			dispatch(setLoadingError(error))
		} finally {
			dispatch(serachLoadingEnd())
		}
	}

export const getUsersList = () => (state: RootState) => {
	return state.search.entities
}

export const getUsersCount = () => (state: RootState) => {
	return state.search.usersCount
}

export const getSearchLoadingStatus = () => (state: RootState) => {
	return state.search.isLoading
}

export const getSearchCurrentPage = () => (state: RootState) => {
	return state.search.page
}

export default searchSlice.reducer
