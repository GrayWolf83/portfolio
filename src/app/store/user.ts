import { IRepo } from '../types/IRepo'
import { AppDispatch, RootState } from './index'
import { setLoadingError } from './error'
import { createSlice, PayloadAction } from '@reduxjs/toolkit'
import { IUser } from '../types/IUser'
import userService from '../services/user.service'

type userState = {
	currentUser: IUser | null
	isLoading: boolean
	error: string | null
	dataLoaded: boolean
	repos: IRepo[]
	skills: string[]
}

const initialState: userState = {
	currentUser: null,
	isLoading: false,
	error: null,
	dataLoaded: false,
	repos: [],
	skills: [],
}

const userSlice = createSlice({
	name: 'user',
	initialState,
	reducers: {
		userLoadingStart: (state) => {
			state.isLoading = true
		},
		userLoadingEnd: (state) => {
			state.isLoading = false
		},
		userLoadingError: (state, action: PayloadAction<string>) => {
			state.error = action.payload
		},
		userLoaded: (state, action: PayloadAction<IUser>) => {
			state.currentUser = action.payload
			state.dataLoaded = true
		},
		userReposLoaded: (state, action: PayloadAction<IRepo[]>) => {
			state.repos = action.payload
		},
		userSkillsLoaded: (
			state,
			action: PayloadAction<{ [key: string]: number }[]>,
		) => {
			let skills = action.payload
				.map((item) => {
					return Object.keys(item).map((name) => name)
				})
				.flat()

			state.skills = Array.from(new Set(skills))
		},
	},
})

const {
	userLoadingStart,
	userLoadingEnd,
	userLoaded,
	userReposLoaded,
	userSkillsLoaded,
	userLoadingError,
} = userSlice.actions

export const loadUser = (login: string) => async (dispatch: AppDispatch) => {
	dispatch(userLoadingStart())
	try {
		const payload = await userService.getUser(login)
		dispatch(userLoaded(payload))

		const repos = await userService.getRepos(login)
		dispatch(userReposLoaded(repos))
	} catch (error: any) {
		dispatch(userLoadingError(error))
	} finally {
		dispatch(userLoadingEnd())
	}
}

export const loadUserSkills =
	(names: string[]) => async (dispatch: AppDispatch) => {
		dispatch(userLoadingStart())
		try {
			const skills = await userService.getSkills(names)
			dispatch(userSkillsLoaded(skills))
		} catch (error: any) {
			dispatch(userLoadingError(error))
		} finally {
			dispatch(userLoadingEnd())
		}
	}

export const getCurrentUser = () => (state: RootState) => {
	return state.user.currentUser
}

export const getUserRepos = () => (state: RootState) => {
	return state.user.repos
}

export const getUserSkills = () => (state: RootState) => {
	return state.user.skills
}

export const getUserLoadingStatus = () => (state: RootState) => {
	return state.user.isLoading
}

export default userSlice.reducer
