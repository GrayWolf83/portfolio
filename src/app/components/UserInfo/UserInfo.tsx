import React from 'react'
import {
	Avatar,
	Box,
	Button,
	Card,
	CardActions,
	CardContent,
	Grid,
	Link,
	Typography,
} from '@mui/material'
import { useAppSelector } from '../../hooks/useAppReduxHooks'
import { getUserRepos, getUserSkills } from '../../store/user'
import { IUser } from '../../types/IUser'

interface IUserInfo {
	user: IUser
}

const UserInfo: React.FC<IUserInfo> = ({ user }) => {
	const skills = useAppSelector(getUserSkills())
	const repos = useAppSelector(getUserRepos())

	return (
		<Grid container spacing={4} data-testid='UserInfo'>
			<Grid item md={4} xs={12} padding='10px'>
				<Card sx={{ height: '100%', p: '10px' }}>
					<Avatar
						src={user.avatar_url}
						sx={{ width: '150px', height: '150px', m: '10px auto' }}
					/>
					<Typography variant='h5'>{user.login}</Typography>
					<Typography variant='h6'>{user.name}</Typography>
					{user.location && (
						<Typography variant='body1'>
							Локация: {user.location}
						</Typography>
					)}
					<Typography variant='body1'>
						Аккаунт создан:{' '}
						{new Date(user.created_at).toLocaleDateString()}
					</Typography>
					{user.blog && (
						<Typography variant='body1'>
							Блог:{' '}
							<Link href={user.blog} target='_blank'>
								{user.blog}
							</Link>
						</Typography>
					)}
				</Card>
			</Grid>
			<Grid item md={8} xs={12} padding='10px'>
				<Card sx={{ height: '100%', p: '10px' }}>
					<Typography variant='h6'>Технологии:</Typography>
					<Grid container spacing={2}>
						{skills.map((skill) => (
							<Grid
								item
								sm={3}
								xs={6}
								key={skill}
								sx={{
									width: '100%',
									mt: '20px',
								}}>
								<Box
									sx={{
										width: '90px',
										height: '90px',
										border: '10px solid var(--primary-color)',
										borderRadius: '50%',
										display: 'flex',
										flexDirection: 'column',
										alignItems: 'center',
										justifyContent: 'center',
										margin: '0 auto',
									}}>
									<Typography
										variant='body1'
										textAlign='center'>
										{skill}
									</Typography>
								</Box>
							</Grid>
						))}
					</Grid>
				</Card>
			</Grid>
			<Grid item xs={12} padding='10px'>
				<Card sx={{ width: '100%', p: '10px' }}>
					<Typography variant='h6'>Проекты:</Typography>
					<Grid container spacing={2}>
						{repos.length &&
							repos.map((repo) => (
								<Grid item key={repo.name} xs={12} sm={4}>
									<Card
										elevation={3}
										sx={{
											height: '100%',
											display: 'flex',
											flexDirection: 'column',
											justifyContent: 'space-between',
										}}>
										<CardContent>
											<Typography variant='h5'>
												{repo.name}
											</Typography>
											{repo.description && (
												<Typography variant='body2'>
													{repo.description}
												</Typography>
											)}
											<Typography variant='body1'>
												ЯП: {repo.language}
											</Typography>
											<Typography variant='body1'>
												Создан:{' '}
												{new Date(
													repo.created_at,
												).toLocaleDateString()}
											</Typography>
										</CardContent>
										<CardActions
											sx={{
												display: 'flex',
												justifyContent: 'space-between',
											}}>
											<Button>Подробнее</Button>
											{repo.homepage && (
												<Link
													href={repo.homepage}
													target='_blank'>
													<Button>
														Ссылка на проект
													</Button>
												</Link>
											)}
										</CardActions>
									</Card>
								</Grid>
							))}
					</Grid>
				</Card>
			</Grid>
		</Grid>
	)
}

export default UserInfo
