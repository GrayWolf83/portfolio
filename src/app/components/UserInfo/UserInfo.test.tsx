import React from 'react'
import { render, screen } from '@testing-library/react'
import { BrowserRouter } from 'react-router-dom'
import { Provider } from 'react-redux'
import store from '../../store'
import UserInfo from './UserInfo'
import { IUser } from '../../types/IUser'

const user: IUser = {
	name: 'a',
	avatar_url: 'aaa',
	login: 'aa',
	html_url: 'ssss',
	blog: 'ddd',
	location: 'Rus',
	created_at: String(new Date()),
	public_repos: 5,
}

describe('UserInfo component', () => {
	test('UserInfo render', () => {
		render(
			<Provider store={store}>
				<BrowserRouter>
					<UserInfo user={user} />
				</BrowserRouter>
			</Provider>,
		)

		const userInfo = screen.getByTestId('UserInfo')
		expect(userInfo).toBeInTheDocument()
	})

	test('UserInfo snapshot', () => {
		expect(
			render(
				<Provider store={store}>
					<BrowserRouter>
						<UserInfo user={user} />
					</BrowserRouter>
				</Provider>,
			),
		).toMatchSnapshot()
	})
})
