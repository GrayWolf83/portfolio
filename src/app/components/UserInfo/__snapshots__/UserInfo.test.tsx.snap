// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserInfo component UserInfo snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-4 css-480o17-MuiGrid-root"
        data-testid="UserInfo"
      >
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-4 css-124xs4j-MuiGrid-root"
        >
          <div
            class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-1o1f417-MuiPaper-root-MuiCard-root"
          >
            <div
              class="MuiAvatar-root MuiAvatar-circular css-b0s0xk-MuiAvatar-root"
            >
              <img
                class="MuiAvatar-img css-1pqm26d-MuiAvatar-img"
                src="aaa"
              />
            </div>
            <h5
              class="MuiTypography-root MuiTypography-h5 css-ag7rrr-MuiTypography-root"
            >
              aa
            </h5>
            <h6
              class="MuiTypography-root MuiTypography-h6 css-2ulfj5-MuiTypography-root"
            >
              a
            </h6>
            <p
              class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
            >
              Локация: 
              Rus
            </p>
            <p
              class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
            >
              Аккаунт создан:
               
              06.09.2022
            </p>
            <p
              class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
            >
              Блог:
               
              <a
                class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-1ps4owl-MuiTypography-root-MuiLink-root"
                href="ddd"
                target="_blank"
              >
                ddd
              </a>
            </p>
          </div>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-8 css-1a57np-MuiGrid-root"
        >
          <div
            class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-1o1f417-MuiPaper-root-MuiCard-root"
          >
            <h6
              class="MuiTypography-root MuiTypography-h6 css-2ulfj5-MuiTypography-root"
            >
              Технологии:
            </h6>
            <div
              class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 css-mhc70k-MuiGrid-root"
            />
          </div>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-z5nknu-MuiGrid-root"
        >
          <div
            class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-kaz5jg-MuiPaper-root-MuiCard-root"
          >
            <h6
              class="MuiTypography-root MuiTypography-h6 css-2ulfj5-MuiTypography-root"
            >
              Проекты:
            </h6>
            <div
              class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 css-mhc70k-MuiGrid-root"
            >
              0
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-4 css-480o17-MuiGrid-root"
      data-testid="UserInfo"
    >
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-4 css-124xs4j-MuiGrid-root"
      >
        <div
          class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-1o1f417-MuiPaper-root-MuiCard-root"
        >
          <div
            class="MuiAvatar-root MuiAvatar-circular css-b0s0xk-MuiAvatar-root"
          >
            <img
              class="MuiAvatar-img css-1pqm26d-MuiAvatar-img"
              src="aaa"
            />
          </div>
          <h5
            class="MuiTypography-root MuiTypography-h5 css-ag7rrr-MuiTypography-root"
          >
            aa
          </h5>
          <h6
            class="MuiTypography-root MuiTypography-h6 css-2ulfj5-MuiTypography-root"
          >
            a
          </h6>
          <p
            class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
          >
            Локация: 
            Rus
          </p>
          <p
            class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
          >
            Аккаунт создан:
             
            06.09.2022
          </p>
          <p
            class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
          >
            Блог:
             
            <a
              class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-1ps4owl-MuiTypography-root-MuiLink-root"
              href="ddd"
              target="_blank"
            >
              ddd
            </a>
          </p>
        </div>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-8 css-1a57np-MuiGrid-root"
      >
        <div
          class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-1o1f417-MuiPaper-root-MuiCard-root"
        >
          <h6
            class="MuiTypography-root MuiTypography-h6 css-2ulfj5-MuiTypography-root"
          >
            Технологии:
          </h6>
          <div
            class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 css-mhc70k-MuiGrid-root"
          />
        </div>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-z5nknu-MuiGrid-root"
      >
        <div
          class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-kaz5jg-MuiPaper-root-MuiCard-root"
        >
          <h6
            class="MuiTypography-root MuiTypography-h6 css-2ulfj5-MuiTypography-root"
          >
            Проекты:
          </h6>
          <div
            class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 css-mhc70k-MuiGrid-root"
          >
            0
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
